# Domain Architecture Network Syntax (DANSy)

This is an analysis that applies the linguistic technique n-gram analysis with network theory to protein domain architectures to describe the functional connections between proteins and the molecular systems they are a part of. 

![Overview of the general workflow](Figures/N%20gram%20network%20workflow.png)

How to cite: Please cite our [bioRxiv paper](https://doi.org/10.1101/2024.12.04.626803), which contains further details and specific applications of the code provided here.

Documentation: Coming soon

## Getting started

Create a virtual environment containing all the dependencies for the analysis using the following code in a terminal.

    conda create env -f dansy.yml

Activate the environment using `conda activate dansy` for specific scripts or select the dansy kernel for jupyter notebooks.

Several of the analysis provided here rely on reference files generated by [CoDIAC](https://github.com/NaegleLab/CoDIAC). We have provided the reference files for the analysis conducted in our manuscript here, but do recommend downloading and generating reference files from scratch. To do this please download the SwissProt ID list from [Gencode](https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_48/gencode.v48.metadata.SwissProt.gz), and run the following code in a terminal. (Note: This will take a bit of time after a fresh install ~30-45min.)

    conda create env -f codiac.yml
    conda activate codiac-env
    python scripts/whole_proteome_reference.py
    conda deactivate

This will create a new directory containing the complete proteome reference file.

For more specific analysis, the following datasets are recommended to be downloaded.

| Analysis | Dataset Source/Code |
| -------- | ------------------- |
| Fusion Gene Analysis|ChimerSeq Excel File from [ChimerDB](https://www.kobic.re.kr/chimerdb/download)|
| TCGA Clinical Data | [cBioPortal](https://www.cbioportal.org/)
| PTM Systems| Use CoDIAC see example Jupyter Notebook|

## How to use this code

To run DANSy on a set of proteins of interest:
    
    import ngramNets
    import generateCompleteProteome
    import pandas as pd

    # Generate the reference proteome dataframe to save time.
    ref_df, _ = generateCompleteProteome.import_proteome_files()
    
    # Generating the DANSy object
    my_dansy = ngramNets.dansy(ref=ref_df, protsOI = my_proteins, n = 10)

The DANSy object can then be used for downstream analysis and has several built-in methods to help. These include:

    # To show the network
    my_dansy.draw_network()

    # To get a summary of the network and n-grams
    my_dansy.summary()

    # To get specific information on protein(s) of interest
    my_dansy.retrieve_protein_info(prot = uniprot_ids_of_interest)

    # To get information on the proteins containing a specific n-gram
    my_dansy.retrieve_protein_info(ngram = ngram_of_interst)

Further, any analysis that can be done on a networkx Graph can be performed on the dansy object by calling the G attribute as shown below

    import networkx as nx

    nx.number_of_nodes(my_dansy.G)
    nx.spring_layout(my_dansy.G)

A separate class for differential gene expression is in development, but can be previewed in the [deg_dansy_preview.ipynb](deg_dansy_preview.ipynb)

## What else is included here

We have provided several Jupyter notebooks that serve as examples of this analysis. Below are short summaries of their applications and results.

### Complete human proteome analysis

Analyzing the domain architectures across the human proteome and broadly characterizing the resulting network and information encoded by different versions related to n-gram length. Here, we find limiting n-grams to 10-grams in length will recapitulate network characteristics of the complete proteome and that specific domains such as the protein kinase, zinc finger C2H2, and EGF-like domains are n-grams that frequently lie along the shortest path and are connected to the most other n-gram nodes in the network.

Associated notebooks:

- [human_model_comp.ipynb](human_model_comp.ipynb)
- [human_proteome.ipynb](human_proteome.ipynb)

### PTM System Analysis

Focused on reversible post-translational modification systems (e.g. phosphorylation, methylation, acetylation) that operate under a reader-writer-eraser paradigm. Characterizing broad properties of how individual components combine in domain architectures and identifying general grammatical rules where eraser domains do not require additional reader domains to modify their activity. Meanwhile, reader domains will frequently associate as "adverbs" to modify the activity of writer domains. Further, the catalytic domains that act as "verbs" must be on separate proteins.

Associated notebooks:

- [PTM_systems_analysis.ipynb](PTM_systems_analysis.ipynb)

### PTM System Evolution

Additional analysis on the phosphorylation systems that compares the network characteristics of domains associated with phosphotyrosine (pTyr) and phosphoserine/threonine (pSer/Thr) systems during the evolutionary period from yeast to humans. The pTyr system is evolutionarily younger and rapidly expanded during the transition to metazoans. Our analysis suggests that during this transition, species were sampling several configurations of the network before converging to a similar set of grammatical rules observed in other PTM systems.

Associated notebooks:

- [ngram_species_comparison.ipynb](ngram_species_comparison.ipynb)

### Fusion Gene Analysis

Here, we analyze how fusion genes may provide an avenue to explore new grammatical structures of domain architectures. Using network topology, we studied fusion genes reported in TCGA and find that the domain architectures of gene fusions still adhere to the same grammatical rules that were established for domain architectures within the complete proteome. We further explore kinase fusions and find that the diversity of domain combinations it is involved in both within the natural proteome and fusions reflect its involvement across biological functions by connecting distinct n-gram families.

Associated notebooks:

- [fusion_genes.ipynb](fusion_genes.ipynb)
- [fusion_gene_survival.ipynb](fusion_gene_survival.ipynb)

### Applications to Differential Gene Expression Analysis

Coming soon, but a preview is in [deg_dansy_preview.ipynb](deg_dansy_preview.ipynb). 

This analysis in the preview uses a differential gene expression dataset from the paper [Convergent approaches to delineate the metabolic regulation of tumor invasion by hyaluronic acid biosynthesis by Shimpi AA, Tan ML, et al.](https://doi.org/10.1002/adhm.202202224) with [PMID: 36479976](https://pubmed.ncbi.nlm.nih.gov/36479976/).